cmd/namespace-lock.go

```diff
@@ -101,42 +99,34 @@ func (n *nsLockMap) lock(ctx context.Context, volume string, path string, lockSo
 	if !locked { // We failed to get the lock
 
 		// Decrement ref count since we failed to get the lock
-		n.lockMapMutex.Lock()
-		nsLk.ref--
-		if nsLk.ref == 0 {
+		if atomic.AddUint32(&nsLk.ref, ^uint32(0)) == 0 {
 			// Remove from the map if there are no more references.
+			n.lockMapMutex.Lock()
 			delete(n.lockMap, resource)
+			n.lockMapMutex.Unlock()
```

Comment:
@harshavardhana Should the ref count be re-checked after the lock is acquired? It looks as though the count could be incremented between the check on line 102 and lock acquisition.

Reply:
it is implicit, the ref count has increased always before the lock is given - if the lock failed i.e based on the timeout then the ref count decreases. 

if the lock is given nothing needs to be done @vcabbage

Reply:
I meant that it's possible that another goroutine could increment the ref between line 102 and getting the `n.lockMapMutex.Lock()`, meaning that `resource` would incorrectly be removed from `n.lockMap`.

It's possible that I'm missing something though. Thanks for the response!

Reply:
it won't be, any new value should be reflected in the ref if we expect the atomic to work. what you are saying was the case previously though where the atomic was not present to serialize such a situation. So whoever gets to increment the `ref` count will always reflect what is present in *nsLk..

Reply:
Unless if you have reproduced a situation, which we need to debug I don't see why it would occur with present restrictions in place.

Reply:
@harshavardhana Since it's dependent on how goroutines are scheduled, it's not easy to reproduce.

This test usually fails after 30-40 seconds on my system, but sometimes it happens faster and sometimes it takes several minutes.

```go
func TestNSLockRace(t *testing.T) {
	ctx := context.Background()

	for i := 0; i < 100000; i++ {
		nsLk := newNSLock(false)

		// lk1; ref=1
		if !nsLk.lock(ctx, "volume", "path", "source", "opsID", false, time.Second) {
			t.Fatal("failed to aqcuire lock")
		}

		// lk2
		lk2ch := make(chan struct{})
		go func() {
			defer close(lk2ch)
			nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 1*time.Millisecond)
		}()
		time.Sleep(1 * time.Millisecond) //wait for goroutine to advance to L96; ref=2

		// Unlock the 1st lock; ref=1 after this line
		nsLk.unlock("volume", "path", false)

		// To trigger the race, lk2 needs to get past L102 (fail to acquire and check ref count),
		// but not yet execute L104 (acquire lockMapMutex)
		//
		// Taking another lockMapMutex here allows queuing up additonal lockers. This should
		// not be required but makes reproduction much easier.
		nsLk.lockMapMutex.Lock()

		// lk3, blocks at L79
		lk3ch := make(chan bool)
		go func() {
			lk3ch <- nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 0)
		}()

		// lk4, blocks at L79
		lk4ch := make(chan bool)
		go func() {
			lk4ch <- nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 0)
		}()
		runtime.Gosched()

		// unlock the manual lock
		nsLk.lockMapMutex.Unlock()

		// To trigger the race:
		// 1) lk3 or lk4 need to advance and increment the ref on the existing resource,
		//    successfully acquiring the lock.
		// 2) lk2 then needs to advance and remove the resource from lockMap.
		// 3) lk3 or lk4 (whichever didn't execute in step 1) then executes and creates
		//    a new entry in lockMap and acquires a lock for the same resource.

		<-lk2ch
		lk3ok := <-lk3ch
		lk4ok := <-lk4ch

		if lk3ok && lk4ok {
			t.Fatalf("multiple locks acquired; iteration=%d, lk3=%t, lk4=%t", i, lk3ok, lk4ok)
		}
	}
}
```

Reply:
Thanks will take a look..

Reply:
> ```go
> func TestNSLockRace(t *testing.T) {
> 	ctx := context.Background()
> 
> 	for i := 0; i < 100000; i++ {
> 		nsLk := newNSLock(false)
> ```

@vcabbage the race seems to exist because of prior ref count update without making sure the lock was indeed acquired, which shouldn't happen in the first place. and it should be only an unlocked caller that should remove the map entry.

Making such simplification fixes the problem and thanks a bunch for the reproducer - its quite hard to find.